client/consensus/grandpa/finality_proof_test.go

```diff
@@ -0,0 +1,506 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+	"testing"
+
+	finalityGrandpa "github.com/ChainSafe/gossamer/pkg/finality-grandpa"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/constraints"
+)
+
+// Check GRANDPA proof-of-finality for the given block.
+//
+// Returns the vector of headers that MUST be validated + imported
+// AND if at least one of those headers is invalid, all other MUST be considered invalid.
+func checkFinalityProof[
+	Hash constraints.Ordered,
+	N constraints.Unsigned,
+	S comparable,
+	H Header[Hash, N],
+	ID AuthorityID,
+](
+	t *testing.T,
+	currentSetID uint64,
+	currentAuthorities AuthorityList[ID],
+	remoteProof []byte) (FinalityProof[Hash, N, H], error) {
+	t.Helper()
+	proof := FinalityProof[Hash, N, H]{}
+	err := scale.Unmarshal(remoteProof, &proof)
+	if err != nil {
+		return FinalityProof[Hash, N, H]{}, fmt.Errorf("failed to decode finality proof %s", err)
```

Comment:
I don't think you need to return an error from a helper function.  Probably makes more sense to just `t.Error()` or `t.Fail()`

Reply:
Yeah because the logic is repeated (but sometimes failing at different parts), but if this is confusing I can change it

Reply:
well given that you never use the `t` parameter, it's not really a helper function.